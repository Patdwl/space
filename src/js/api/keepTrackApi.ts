import { SatMath } from '../satMath/satMath';
import { KeepTrackApi, SatObject, SettingsManager } from './keepTrackTypes';
import { html } from './templateLiterals';

declare global {
  // eslint-disable-next-line no-unused-vars
  interface Window {
    settingsManager: SettingsManager;
    jQuery: unknown;
    $: unknown;
    gremlins: any;
    randomizer: any;
    keepTrackApi: KeepTrackApi;
    dataLayer: any; // For Google Tag Manager
    _numeric: any;
    satellite: SatMath;
    M: any;
  }
}

export const register = (params: { method: string; cbName: string; cb: any }) => {
  // If this is a valid callback
  if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
    // Add the callback
    keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
  } else {
    throw new Error(`Invalid callback "${params.method}"!`);
  }
};
export const unregister = (params: { method: string; cbName: string }) => {
  // If this is a valid callback
  if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
    for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
      if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
        keepTrackApi.callbacks[params.method].splice(i, 1);
        return;
      }
    }
    // If we got this far, it means we couldn't find the callback
    throw new Error(`Callback "${params.cbName} not found"!`);
  } else {
    // Couldn't find the method
    throw new Error(`Invalid callback "${params.method}"!`);
  }
};
export const isThisJest = () => typeof process !== 'undefined';

export const keepTrackApi: KeepTrackApi = {
  html: html,
  register: register,
  unregister: unregister,
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
    onCruncherReady: [],
    onCruncherMessage: [],
    uiManagerInit: [],
    uiManagerOnReady: [],
    bottomMenuClick: [],
    hideSideMenus: [],
    nightToggle: [],
    orbitManagerInit: [],
    adviceReady: [],
    drawManagerLoadScene: [],
    drawOptionalScenery: [],
    updateLoop: [],
    rmbMenuActions: [],
    rightBtnMenuAdd: [],
    updateDateTime: [],
    uiManagerFinal: [],
    loadCatalog: [],
    resetSensor: [],
    setSensor: [],
    onDrawLoopComplete: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      keepTrackApi.programs.soundManager.play('whoosh');
      keepTrackApi.callbacks.selectSatData.forEach((cb: any) => cb.cb(sat, satId));
    },
    updateSelectBox: (sat: any) => {
      keepTrackApi.callbacks.updateSelectBox.forEach((cb: any) => cb.cb(sat));
    },
    onCruncherReady: () => {
      keepTrackApi.callbacks.onCruncherReady.forEach((cb: any) => cb.cb());
    },
    onCruncherMessage: () => {
      keepTrackApi.callbacks.onCruncherMessage.forEach((cb: any) => cb.cb());
    },
    uiManagerInit: () => {
      keepTrackApi.callbacks.uiManagerInit.forEach((cb: any) => cb.cb());
    },
    uiManagerOnReady: () => {
      keepTrackApi.callbacks.uiManagerOnReady.forEach((cb: any) => cb.cb());
    },
    bottomMenuClick: (iconName: string) => {
      keepTrackApi.programs.soundManager.play('genericBeep');
      keepTrackApi.callbacks.bottomMenuClick.forEach((cb: any) => cb.cb(iconName));
    },
    hideSideMenus: () => {
      keepTrackApi.callbacks.hideSideMenus.forEach((cb: any) => cb.cb());
    },
    nightToggle: (gl: any, nightTexture: any, texture: any) => {
      keepTrackApi.callbacks.nightToggle.forEach((cb: any) => cb.cb(gl, nightTexture, texture));
    },
    orbitManagerInit: () => {
      keepTrackApi.callbacks.orbitManagerInit.forEach((cb: any) => cb.cb());
    },
    adviceReady: () => {
      keepTrackApi.callbacks.adviceReady.forEach((cb: any) => cb.cb());
    },
    drawManagerLoadScene: () => {
      keepTrackApi.callbacks.drawManagerLoadScene.forEach((cb: any) => cb.cb());
    },
    drawOptionalScenery: () => {
      keepTrackApi.callbacks.drawOptionalScenery.forEach((cb: any) => cb.cb());
    },
    updateLoop: () => {
      keepTrackApi.callbacks.updateLoop.forEach((cb: any) => cb.cb());
    },
    rmbMenuActions: (menuName: string, satnum = -1) => {
      keepTrackApi.callbacks.rmbMenuActions.forEach((cb: any) => cb.cb(menuName, satnum));
    },
    rightBtnMenuAdd: () => {
      keepTrackApi.callbacks.rightBtnMenuAdd.forEach((cb: any) => cb.cb());
    },
    updateDateTime: (date: Date) => {
      keepTrackApi.callbacks.updateDateTime.forEach((cb: any) => cb.cb(date));
    },
    uiManagerFinal: () => {
      keepTrackApi.callbacks.uiManagerFinal.forEach((cb: any) => cb.cb());
    },
    loadCatalog: async (): Promise<SatObject[]> => keepTrackApi.callbacks.loadCatalog[0].cb(),
    resetSensor: () => {
      keepTrackApi.callbacks.resetSensor.forEach((cb: any) => cb.cb());
    },
    setSensor: (sensor: any, id: number) => {
      keepTrackApi.callbacks.setSensor.forEach((cb: any) => cb.cb(sensor, id));
    },
    onDrawLoopComplete: () => {
      keepTrackApi.callbacks.onDrawLoopComplete.forEach((cb: any) => cb.cb());
    },
  },
  programs: <any>{},
};

// First time we call this module we should make it available to the rest of the application
if (!window.keepTrackApi) window.keepTrackApi = keepTrackApi;
