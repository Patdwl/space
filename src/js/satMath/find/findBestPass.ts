import { DEG2RAD, MINUTES_PER_DAY, TAU } from '@app/js/lib/constants';
import { calcSatrec } from '@app/js/satSet/catalogSupport/calcSatrec';
import { SatelliteRecord } from 'ootk';
import { keepTrackApi } from '../../api/keepTrackApi';
import { lookanglesRow, SatObject, SensorObject } from '../../api/keepTrackTypes';
import { getRae } from '../calc/getRae';
import { checkIsInView } from '../lookangles/checkIsInView';

export const findBestPass = (sat: SatObject, sensors: SensorObject[]): lookanglesRow[] => {
  const { timeManager, uiManager } = keepTrackApi.programs;

  // Check if there is a sensor
  if (sensors.length <= 0 || !sensors[0] || typeof sensors[0].obsminaz == 'undefined') {
    uiManager.toast(`Sensor's format incorrect. Did you select a sensor first?`, 'critical');
    return [];
  }
  sensors[0].observerGd = {
    // Array to calculate look angles in propagate()
    lat: sensors[0].lat * DEG2RAD,
    lon: sensors[0].lon * DEG2RAD,
    alt: sensors[0].alt,
  };

  // TOOD: Instead of doing the first sensor this should return an array of TEARRs for all sensors.
  const sensor = sensors[0];

  const simulationTime = timeManager.simulationTimeObj;
  let offset = 0;

  var satrec = calcSatrec(sat);
  var lookanglesTable = []; // Iniially no rows to the table

  let looksInterval = 5;
  let looksLength = 7;

  // Setup flags for passes
  let score = 0;
  let sAz = <string | null>null;
  let sEl = <string | null>null;
  let srng = <string | null>null;
  let sTime = <Date | null>null;
  let passMinrng = sensor.obsmaxrange; // This is set each look to find minimum rng (start at max rng)
  let passMaxEl = 0;
  let start3 = false;
  let stop3 = false;

  let orbitalPeriod = MINUTES_PER_DAY / ((satrec.no * MINUTES_PER_DAY) / TAU); // Seconds in a day divided by mean motion

  const _propagateBestPass = (now: Date, satrecIn: SatelliteRecord): lookanglesRow => {
    let aer = getRae(now, satrecIn, sensor);
    let isInFOV = checkIsInView(sensor, aer);

    if (isInFOV) {
      // Previous Pass to Calculate first line of coverage
      const now1 = timeManager.getOffsetTimeObj(offset - looksInterval * 1000, simulationTime);
      let aer1 = getRae(now1, satrecIn, sensor);

      let isInFOV1 = checkIsInView(sensor, aer1);
      if (!isInFOV1) {
        // if it starts around 3
        if (aer.el <= 3.5) {
          start3 = true;
        }

        // First Line of Coverage
        sTime = now;
        sAz = aer.az.toFixed(0);
        sEl = aer.el.toFixed(1);
        srng = aer.rng.toFixed(0);
      } else {
        // Next Pass to Calculate Last line of coverage
        let _now1 = timeManager.getOffsetTimeObj(offset + looksInterval * 1000, simulationTime);
        aer1 = getRae(_now1, satrecIn, sensor);

        isInFOV1 = checkIsInView(sensor, aer1);
        if (!isInFOV1) {
          // if it stops around 3
          stop3 = aer.el <= 3.5;

          // Skip pass if satellite is in track right now
          if (sTime == null) {
            return {
              sortTime: null,
              scc: null,
              score: null,
              startDate: null,
              startTime: null,
              startAz: null,
              startEl: null,
              startrng: null,
              stopDate: null,
              stopTime: null,
              stopAz: null,
              stopEl: null,
              stoprng: null,
              tic: null,
              minrng: null,
              passMaxEl: null,
            };
          }

          score = Math.min((((now.getTime() - sTime.getTime()) / 1000 / 60) * 10) / 8, 10); // 8 minute pass is max score
          let elScore = Math.min((passMaxEl / 50) * 10, 10); // 50 el or above is max score

          // elScore -= Math.max((passMaxEl - 50) / 5, 0); // subtract points for being over 50 el
          elScore *= start3 && stop3 ? 2 : 1; // Double points for start and stop at 3
          score += elScore;
          score += Math.min((10 * 750) / passMinrng, 10); // 750 or less is max score

          // score -= Math.max((750 - passMinrng) / 10, 0); // subtract points for being closer than 750
          let tic = 0;
          tic = (now.getTime() - sTime.getTime()) / 1000 || 0;

          // Last Line of Coverage
          return {
            sortTime: sTime.getTime(),
            scc: satrecIn.satnum,
            score: score,
            startDate: sTime,
            startTime: sTime,
            startAz: sAz,
            startEl: sEl,
            startrng: srng,
            stopDate: now,
            stopTime: now,
            stopAz: aer.az.toFixed(0),
            stopEl: aer.el.toFixed(1),
            stoprng: aer.rng.toFixed(0),
            tic: tic,
            minrng: passMinrng.toFixed(0),
            passMaxEl: passMaxEl.toFixed(1),
          };
        }
      }
      // Do this for any pass in coverage
      if (passMaxEl < aer.el) passMaxEl = aer.el;
      if (passMinrng > aer.rng) passMinrng = aer.rng;
    }
    return {
      sortTime: null,
      scc: null,
      score: null,
      startDate: null,
      startTime: null,
      startAz: null,
      startEl: null,
      startrng: null,
      stopDate: null,
      stopTime: null,
      stopAz: null,
      stopEl: null,
      stoprng: null,
      tic: null,
      minrng: null,
      passMaxEl: null,
    };
  };

  for (let i = 0; i < looksLength * 24 * 60 * 60; i += looksInterval) {
    // lookanglesInterval in seconds
    offset = i * 1000; // Offset in seconds (msec * 1000)
    const now = timeManager.getOffsetTimeObj(offset, simulationTime);
    if (lookanglesTable.length <= 5000) {
      // Maximum of 1500 lines in the look angles table
      const _lookanglesRow = _propagateBestPass(now, satrec);
      // If data came back...
      if (_lookanglesRow.score !== null) {
        lookanglesTable.push(_lookanglesRow); // Update the table with looks for this 5 second chunk and then increase table counter by 1

        // Reset flags for next pass
        score = 0;
        sAz = null;
        sEl = null;
        srng = null;
        sTime = null;
        passMinrng = sensor.obsmaxrange; // This is set each look to find minimum rng
        passMaxEl = 0;
        start3 = false;
        stop3 = false;
        // Jump 3/4th to the next orbit
        i = i + orbitalPeriod * 60 * 0.75; // NOSONAR
      }
    }
  }

  return lookanglesTable;
};
