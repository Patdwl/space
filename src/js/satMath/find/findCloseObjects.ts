import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { EciPos, SatObject } from '@app/js/api/keepTrackTypes';
import { getUnique } from '@app/js/lib/helpers';
import { getEci } from '../calc/getEci';

let searchStrCache: string;
export const findCloseObjects = () => {
  if (searchStrCache) return searchStrCache;
  const { satSet } = keepTrackApi.programs;
  const searchRadius = 50; // km

  let csoList = [];
  let satList = <SatObject[]>[];

  // Loop through all the satellites
  for (let i = 0; i < satSet.orbitalSats; i++) {
    // Get the satellite
    const sat = satSet.getSat(i);
    // Avoid unnecessary errors
    if (typeof sat.TLE1 == 'undefined') continue;
    // Only look at satellites in LEO
    // if (sat.apogee > 5556) continue;
    // Find where the satellite is right now
    if (typeof sat.position === 'undefined') {
      sat.position = <EciPos>getEci(sat, new Date()).position || { x: 0, y: 0, z: 0 };
    }
    // If it fails, skip it
    if (isNaN(sat.position.x) || isNaN(sat.position.y) || isNaN(sat.position.z)) continue;
    if (sat.position === { x: 0, y: 0, z: 0 }) continue;
    // Add the satellite to the list
    satList.push(sat);
  }

  // Remove duplicates
  satList = getUnique(satList);

  // Sort satList by position.x property
  satList.sort((a, b) => a.position.x - b.position.x);

  // Loop through all the satellites with valid positions
  let i = 0;
  const satListLen = satList.length;
  for (i = 0; i < satListLen; i++) {
    const sat1 = satList[i];
    const pos1 = sat1.position;

    // Calculate the area around the satellite
    const posXmin = pos1.x - searchRadius;
    const posXmax = pos1.x + searchRadius;
    const posYmin = pos1.y - searchRadius;
    const posYmax = pos1.y + searchRadius;
    const posZmin = pos1.z - searchRadius;
    const posZmax = pos1.z + searchRadius;

    // Loop through the list again
    let j = 0;
    for (j = Math.max(0, i - 200); j < satListLen; j++) {
      const sat2 = satList[j]; // Get the second satellite
      if (sat1 == sat2) continue; // Skip the same satellite
      const pos2 = sat2.position; // Get the second satellite's position

      // Satellites are in order of x position so once we exceed the maxX, we can stop
      if (pos2.x > posXmax) break;
      // Check to see if the second satellite is in the search area
      if (pos2.x < posXmax && pos2.x > posXmin && pos2.y < posYmax && pos2.y > posYmin && pos2.z < posZmax && pos2.z > posZmin) {
        // Add the second satellite to the list if it is close
        csoList.push({ sat1: sat1, sat2: sat2 });
      }
    }
  }

  let csoListUnique = getUnique(csoList);

  const csoStrArr = []; // Clear CSO List

  // Loop through the possible CSOs
  for (let i = 0; i < csoListUnique.length; i++) {
    // Calculate the first CSO's position 30 minutes later
    let sat = csoListUnique[i].sat1;
    let eci = getEci(sat, new Date(Date.now() + 1000 * 60 * 30));
    if (eci.position === { x: 0, y: 0, z: 0 }) continue;
    csoListUnique[i].sat1.position = eci.position;

    // Calculate the second CSO's position 30 minutes later
    sat = csoListUnique[i].sat2;
    eci = getEci(sat, new Date(Date.now() + 1000 * 60 * 30));
    if (eci.position === { x: 0, y: 0, z: 0 }) continue;
    sat.position = eci.position;
    csoListUnique[i].sat2.position = eci.position;
  }

  // Loop through the CSOs
  for (let i = 0; i < csoListUnique.length; i++) {
    // Check the first CSO
    let sat1 = csoListUnique[i].sat1;
    let pos1 = sat1.position;
    if (typeof pos1 == 'undefined') continue;

    // Calculate the area around the CSO
    let posXmin = pos1.x - searchRadius;
    let posXmax = pos1.x + searchRadius;
    let posYmin = pos1.y - searchRadius;
    let posYmax = pos1.y + searchRadius;
    let posZmin = pos1.z - searchRadius;
    let posZmax = pos1.z + searchRadius;

    // Get the second CSO object
    let sat2 = csoListUnique[i].sat2;
    let pos2 = sat2.position;
    if (typeof pos2 == 'undefined') continue;

    // If it is still in the search area, add it to the list
    if (pos2.x < posXmax && pos2.x > posXmin && pos2.y < posYmax && pos2.y > posYmin && pos2.z < posZmax && pos2.z > posZmin) {
      csoStrArr.push(sat1.sccNum);
      csoStrArr.push(sat2.sccNum);
    }
  }

  // Generate the search string
  const csoListUniqueArr = Array.from(new Set(csoStrArr));
  let searchStr = '';
  for (let i = 0; i < csoListUniqueArr.length; i++) {
    if (i == csoListUniqueArr.length - 1) {
      searchStr += csoListUniqueArr[i];
    } else {
      searchStr += csoListUniqueArr[i] + ',';
    }
  }

  // Dont need to do this math more than once
  searchStrCache = searchStr;
  return searchStr; // csoListUnique;
};
