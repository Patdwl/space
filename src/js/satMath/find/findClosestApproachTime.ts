import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { SatObject } from '@app/js/api/keepTrackTypes';
import { Sgp4 } from 'ootk';
import { getEci } from '../calc/getEci';
import { sat2ric } from '../transforms';

export const findClosestApproachTime = (
  sat1: SatObject,
  sat2: SatObject,
  propLength?: number
): {
  offset: number;
  dist: number;
  ric: { position: [number, number, number]; velocity: [number, number, number] };
} => {
  let offset = 0;
  propLength ??= 1440 * 60; // 1 Day
  let minDist = 1000000;
  let result = {
    offset: null,
    dist: null,
    ric: null,
  };

  sat1.satrec = Sgp4.createSatrec(sat1.TLE1, sat1.TLE2); // perform and store sat init calcs
  sat2.satrec = Sgp4.createSatrec(sat2.TLE1, sat2.TLE2); // perform and store sat init calcs

  for (let t = 0; t < propLength; t++) {
    offset = t * 1000; // Offset in seconds (msec * 1000)

    sat1 = <SatObject>{ ...sat1, ...getEci(sat1, new Date(Date.now() + offset)) };
    sat2 = <SatObject>{ ...sat2, ...getEci(sat2, new Date(Date.now() + offset)) };
    const pv = sat2ric(sat1, sat2);
    const dist = Math.sqrt(pv.position[0] ** 2 + pv.position[1] ** 2 + pv.position[2] ** 2);
    if (dist < minDist && !(pv.position[0] === 0 && pv.position[1] === 0 && pv.position[2] === 0)) {
      minDist = dist;
      result = {
        offset,
        dist: dist,
        ric: pv,
      };
    }
  }

  // Go to closest approach time
  keepTrackApi.programs.timeManager.changeStaticOffset(result.offset);
  keepTrackApi.programs.satSet.setColorScheme(settingsManager.currentColorScheme, true);

  return result;
};
