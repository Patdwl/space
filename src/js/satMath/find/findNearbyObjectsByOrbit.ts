import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { SatObject } from '@app/js/api/keepTrackTypes';
import { DEG2RAD } from '@app/js/lib/constants';

export const findNearbyObjectsByOrbit = (sat: SatObject) => {
  const { satData: catalog } = keepTrackApi.programs.satSet;

  const maxPeriod = sat.period * 1.1;
  const minPeriod = sat.period * 0.9;
  const maxInclination = sat.inclination + 10 * DEG2RAD;
  const minInclination = sat.inclination - 10 * DEG2RAD;
  let maxRaan = sat.raan + 10 * DEG2RAD;
  let minRaan = sat.raan - 10 * DEG2RAD;
  if (sat.raan >= 350 * DEG2RAD) {
    maxRaan -= 360 * DEG2RAD;
  }
  if (sat.raan <= 10 * DEG2RAD) {
    minRaan += 360 * DEG2RAD;
  }

  return catalog
    .filter((s) => {
      if (s.static) return false;
      if (s.inclination < minInclination || s.inclination > maxInclination) return false;
      if (sat.raan > 350 * DEG2RAD || sat.raan < 10 * DEG2RAD) {
        if (s.raan > maxRaan && s.raan < minRaan) return false;
      } else {
        if (s.raan < minRaan || s.raan > maxRaan) return false;
      }
      if (s.period < minPeriod || s.period > maxPeriod) return false;
      return true;
    })
    .map((s) => s.id);
};
