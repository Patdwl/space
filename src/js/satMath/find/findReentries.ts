import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { SpaceObjectType } from '@app/js/api/SpaceObjectType';

export const findReentries = (): string => {
  const { satSet } = keepTrackApi.programs;
  const reentries = satSet.satData.filter((sat) => sat.type === SpaceObjectType.PAYLOAD || sat.type === SpaceObjectType.ROCKET_BODY || sat.type === SpaceObjectType.DEBRIS);

  const reentriesStr = reentries
    .filter((sat) => sat.perigee > 0)
    .sort((a, b) => a.perigee - b.perigee)
    .slice(0, 100)
    .map((sat) => sat.sccNum)
    .join(',');

  return reentriesStr;
};
