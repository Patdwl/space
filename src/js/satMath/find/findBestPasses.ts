import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { SensorObject } from '@app/js/api/keepTrackTypes';
import { saveCsv } from '@app/js/lib/helpers';
import { findBestPass } from './findBestPass';

export const findBestPasses = (sats: string, sensor: SensorObject) => {
  const { satSet } = keepTrackApi.programs;

  sats = sats.replace(/ /gu, ',');
  const satArray = sats.split(',');
  let tableSatTimes = [];
  for (let i = 0; i < satArray.length; i++) {
    try {
      const satId = satArray[i];
      if (typeof satId == 'undefined' || satId == null || satId === '' || satId === ' ') continue;
      const sat = satSet.getSatFromObjNum(parseInt(satId));
      const satPasses = findBestPass(sat, [sensor]);
      for (let s = 0; s < satPasses.length; s++) {
        tableSatTimes.push(satPasses[s]);
        // }
      }
    } catch (e) {
      console.debug(e);
    }
  }
  tableSatTimes.sort((a, b) => b.sortTime - a.sortTime);
  tableSatTimes.reverse();
  tableSatTimes.forEach((v) => {
    delete v.sortTime;
  });

  for (let i = 0; i < tableSatTimes.length; i++) {
    tableSatTimes[i].startDate = tableSatTimes[i].startDate.toISOString().split('T')[0];
    tableSatTimes[i].startTime = tableSatTimes[i].startTime.toISOString().split('T')[1].split('.')[0];
    tableSatTimes[i].stopDate = tableSatTimes[i].stopDate.toISOString().split('T')[0];
    tableSatTimes[i].stopTime = tableSatTimes[i].stopTime.toISOString().split('T')[1].split('.')[0];
  }

  saveCsv(tableSatTimes, 'bestSatTimes');
};
