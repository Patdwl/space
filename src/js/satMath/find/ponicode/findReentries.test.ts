import { keepTrackApiStubs } from '@app/js/api/apiMocks';
import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { KeepTrackPrograms } from '@app/js/api/keepTrackTypes';
import * as findReentries from '@app/js/satMath/find/findReentries';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });

// @ponicode
describe('findReentries.findReentries', () => {
  test('0', () => {
    let result: any = findReentries.findReentries();
    expect(result).toMatchSnapshot();
  });
});
