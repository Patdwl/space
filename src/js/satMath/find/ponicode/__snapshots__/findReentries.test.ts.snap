// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findReentries.findReentries 0 1`] = `"00005,00005"`;
